src/System.Linq.Expressions/tests/TypeBinary/TypeBinaryTests.cs

```diff
@@ -44,12 +54,32 @@ private static IEnumerable<Expression> Constants
                 yield return Expression.Constant(1);
                 yield return Expression.Constant(DateTime.MinValue);
                 yield return Expression.Constant("hello");
+                yield return Expression.Constant("hello", typeof(object));
                 yield return Expression.Constant(Expression.Empty());
                 yield return Expression.Constant(Expression.And(Expression.Constant(1), Expression.Constant(2)));
                 yield return Expression.Constant(typeof(int));
                 yield return Expression.New(typeof(string).GetConstructor(new[] { typeof(char[]) }), Expression.Constant(new[] { 't', 'e', 's', 't' }));
                 yield return Expression.Convert(Expression.Constant(DateTime.MaxValue), typeof(DateTime?));
                 yield return Expression.Constant(1, typeof(int?));
+                yield return Expression.Constant(1, typeof(object));
+                yield return Expression.Constant(1, typeof(IComparable));
+                yield return Expression.Constant(DateTime.MinValue, typeof(object));
+                yield return Expression.Constant(1, typeof(ValueType));
+                yield return Expression.Constant(DateTime.MinValue, typeof(ValueType));
+                yield return Expression.Constant(DateTime.MinValue, typeof(IComparable));
+                yield return Expression.Constant(ByteBased.A);
```

Comment:
also 
yield return Expression.Constant(ByteBased.A, typeof(ByteBased?));      

As I remember, interpreter has some fairly arcane code to deal with nullable enums, so I am a bit paranoid :-)